### PR TITLE
Add KeysWithFilter & ListKeysWithFilters methods to KV interface

### PR DESCRIPTION
**Pull request by Darshan174**
implemented a method KeysWithFilters to return keys matching specified filters and ListKeysWithFilters to provide an iterable list of keys (as suggested). And additionally defined KeyIterator for iterating over keys and a utility function matchesFilters to check if a key meets all filter criteria.

### CODE REVIEW

**Comment by Darshan174** [comment]
Closes #1655

File: jetstream/kv.go [hunk 1]
```diff
@@ -484,6 +484,63 @@ func (js *jetStream) CreateKeyValue(ctx context.Context, cfg KeyValueConfig) (Ke
 		return nil, err
 	}
 
+// KeysWithFilters returns a filtered list of keys in the bucket.
+func (js *jetStream) KeysWithFilters(filter []string) ([]string, error) {
```
Comment (Jarema) [inline jetstream/kv.go:488]: Filtering the keys does not mean filtering them on client side code.
It means passing `subject filters` to `ordered consumer`. Take a look how `ListKeys` or `Keys` are implemented.

File: jetstream/kv.go [hunk 2]
```diff
@@ -484,6 +484,63 @@ func (js *jetStream) CreateKeyValue(ctx context.Context, cfg KeyValueConfig) (Ke
...
+    return true
+}
+
+// KeyIterator is a simple iterator for keys.
+type KeyIterator struct {
+    keys []string
+    index int
+}
+
+// NewKeyIterator creates a new KeyIterator.
```
Comment (Jarema) [inline jetstream/kv.go:528]: To see how those APIs should work in nats.go client, please take a look at `ListKeys` method.

**Review by Jarema — CHANGES_REQUESTED** [review]
Hey!

Thanks for your effort.
However, this PR needs quite some work, as it does not do what it should do.

Please take a look at `Keys` or `ListKeys` to see how we are using `ordered consumer` to fetch the keys from the server.

Also, this is where they should be filtered out - by providing `subject filters` with appropriate subjects.

**Comment by Darshan174** [comment]
> Hey!
> 
> Thanks for your effort. However, this PR needs quite some work, as it does not do what it should do.
> 
> Please take a look at `Keys` or `ListKeys` to see how we are using `ordered consumer` to fetch the keys from the server.
> 
> Also, this is where they should be filtered out - by providing `subject filters` with appropriate subjects.

Hey @Jarema! Thank you for your patience in guiding me. I'm having a hard time figuring out the `subject filters` which u suggested. Could you please provide more details on this?

**Comment by Darshan174** [comment]
Hey @Jarema! Thank you for your patience in guiding me. I've read through the [docs](https://docs.nats.io/nats-concepts/jetstream/streams#subjects) about FIlterSubject (I hope this is what you were referring to), but I'd a hard time understanding the implimentation of passing subject filters to ordered customers in ListKeys from the Kv.go file. Could you please provide more details on this?

**Pull request closed without merging**